framework/include/samplers/SamplerInterface.h

```diff
@@ -63,3 +65,21 @@ class SamplerInterface
   /// Thread ID
   THREAD_ID _smi_tid;
 };
+
+template <typename T>
+T &
+SamplerInterface::getSampler(const std::string & name)
+{
+  return getSamplerByName<T>(_smi_params.get<SamplerName>(name));
+}
+
+template <typename T>
+T &
+SamplerInterface::getSamplerByName(const SamplerName & name)
+{
+  Sampler * base_ptr = &_smi_feproblem.getSampler(name, _smi_tid);
+  T * obj_ptr = dynamic_cast<T *>(base_ptr);
+  if (!obj_ptr)
+    mooseError("Failed to find a Sampler object with the name '", name, "' for the desired type.");
+  return *obj_ptr;
```

Comment:
This "ByName" interface doesn't match the others in MOOSE. We typically don't do a dynamic cast. You might take a look and see if you can rework it. Also, you shouldn't need to convert to a pointer to cast.

Reply:
The cast only happens when a template parameter is supplied. By default T = Sampler and there is a specialization in the C file that just returns the base object.

This is meant to mimic the getUserObject call, but by default returns the base object type.

If you want to be able to handle failure then you need to use a pointer, you can't have a null reference.